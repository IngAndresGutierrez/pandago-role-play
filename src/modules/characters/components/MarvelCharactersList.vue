<template>
  <v-app>
    <v-container>
      <v-row class="mb-4" justify="center">
        <v-col
          cols="12"
          md="6"
          sm="10"
        >
          <v-text-field
            v-model="searchCharacter"
            clearable
            hint="Ingresa el nombre del personaje"
            label="Buscar personaje"
            prepend-inner-icon="mdi-magnify"
          />
        </v-col>
        <v-col
          cols="12"
          md="3"
          sm="10"
        >
          <v-select
            v-model="selectedComics"
            clearable
            item-title="title"
            item-value="id"
            :items="comicsOptions"
            label="Comics"
            prepend-inner-icon="mdi-filter"
          />
        </v-col>
        <v-col
          cols="12"
          md="3"
          sm="10"
        >
          <v-select
            v-model="selectedSeries"
            clearable
            item-title="title"
            item-value="id"
            :items="seriesOptions"
            label="Series"
            prepend-inner-icon="mdi-filter"
          />
        </v-col>
      </v-row>
    </v-container>

    <v-container v-if="loading" fluid>
      <v-row>
        <v-col
          v-for="character in dummySkeleton"
          :key="character"
          cols="12"
          lg="3"
          md="4"
          sm="6"
        >
          <v-skeleton-loader class="mx-auto border" max-width="300" type="image, article" />
        </v-col>
      </v-row>
    </v-container>

    <v-row justify="center">
      <v-col cols="10">
        <v-alert v-if="!characters.length && !loading" type="info">No existe informaci√≥n para los filtros seleccionados.</v-alert>
      </v-col>
    </v-row>

    <v-container v-if="characters.length" fluid>
      <v-row>
        <v-col
          v-for="character in characters"
          :key="character.id"
          cols="12"
          lg="3"
          md="4"
          sm="6"
        >
          <MarvelCharacterCard :character="character" />
        </v-col>
      </v-row>
    </v-container>

    <v-overlay
      class="align-center justify-center"
      :model-value="loading"
    >
      <v-progress-circular
        color="primary"
        indeterminate
        size="64"
      />
    </v-overlay>

    <div class="text-center">
      <v-container>
        <v-row justify="center">
          <v-col cols="8">
            <v-container class="max-width">
              <v-pagination
                v-model="currentPage"
                class="my-4"
                :length="totalPages"
              />
            </v-container>
          </v-col>
        </v-row>
      </v-container>
    </div>
  </v-app>
</template>

<script setup lang="ts">
  import { onMounted } from 'vue'
  import debounce from 'lodash/debounce'

  import marvelService from '../services/characters'
  import marvelCommonService from '../../common/services/filters'

  import type { Character } from '../models/characters'
  import MarvelCharacterCard from './MarvelCharacterCard.vue'

  const characters = ref<Character[]>([])
  const dummySkeleton = [1, 2, 3, 4, 5, 6, 7, 8]
  const currentPage = ref(1)
  const totalPages = ref(1)
  const limit = 20
  const loading = ref(false)
  const searchCharacter = ref(null)
  const selectedComics = ref()
  const selectedSeries = ref()
  const seriesOptions = ref([])
  const comicsOptions = ref([])

  const fetchCharacters = async () => {
    loading.value = true

    const offset = (currentPage.value - 1) * limit
    const newCharacters = await marvelService.getCharacters(
      offset,
      limit,
      searchCharacter.value,
      selectedSeries.value,
      selectedComics.value,
    )
    totalPages.value = newCharacters.total
    characters.value = newCharacters.results
    loading.value = false
  }

  const fetchSeries = async () => {
    const newSeries = await marvelCommonService.getSeries()
    seriesOptions.value = newSeries
  }

  const fetchComics = async () => {
    const newComics = await marvelCommonService.getComics()
    comicsOptions.value = newComics
  }

  const initfetchData = () => {
    fetchCharacters()
    fetchComics()
    fetchSeries()
  }

  const debouncedFetchCharacters = debounce(fetchCharacters, 500)

  onMounted(initfetchData)
  watch(searchCharacter, debouncedFetchCharacters)
  watch([currentPage, selectedComics, selectedSeries], fetchCharacters)

</script>
