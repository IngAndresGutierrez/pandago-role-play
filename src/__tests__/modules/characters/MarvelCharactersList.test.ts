import 'vuetify/styles'
import { mount } from '@vue/test-utils'
import { createVuetify } from 'vuetify'
import { describe, expect, test } from 'vitest'

import MarvelCharactersList from '@/modules/characters/components/MarvelCharactersList.vue'

const vuetify = createVuetify()

global.ResizeObserver = require('resize-observer-polyfill')

const mockCharacters = [{
  id: 1,
  name: 'Test character',
  thumbnail: {
    path: 'https://mypath.com/images',
    extension: 'jpg',
  },
}]

describe('MarvelCharactersList.vue', () => {
  test('renders correctly', () => {
    const wrapper = mount(MarvelCharactersList, {
      global: {
        plugins: [vuetify],
      },
      props: {
        character: mockCharacters,
      },
    })

    expect(wrapper.element).toMatchSnapshot()
  })
})
