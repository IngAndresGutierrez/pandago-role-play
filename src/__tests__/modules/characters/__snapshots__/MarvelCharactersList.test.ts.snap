// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MarvelCharactersList.vue > renders correctly 1`] = `
<div
  character="[object Object]"
  class="v-application v-theme--light v-layout v-layout--full-height v-locale--is-ltr"
>
  <div
    class="v-application__wrap"
  >
    
    <div
      class="v-container v-locale--is-ltr"
    >
      <div
        class="v-row justify-center mb-4"
      >
        <div
          class="v-col-sm-10 v-col-md-6 v-col-12"
        >
          <div
            class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-theme--light v-locale--is-ltr v-text-field"
          >
            <!---->
            <div
              class="v-input__control"
            >
              
              <div
                class="v-field v-field--appended v-field--center-affix v-field--prepended v-field--variant-filled v-theme--light v-locale--is-ltr"
              >
                <div
                  class="v-field__overlay"
                />
                <div
                  class="v-field__loader"
                >
                  <div
                    aria-hidden="true"
                    aria-valuemax="100"
                    aria-valuemin="0"
                    class="v-progress-linear v-theme--light v-locale--is-ltr"
                    role="progressbar"
                    style="top: 0px; height: 0px; --v-progress-linear-height: 2px;"
                  >
                    <!---->
                    <div
                      class="v-progress-linear__background"
                    />
                    <div
                      class="v-progress-linear__buffer"
                      style="width: 0%;"
                    />
                    <transition-stub
                      appear="false"
                      css="true"
                      name="fade-transition"
                      persisted="false"
                    >
                      <div
                        class="v-progress-linear__indeterminate"
                      >
                        
                        <div
                          class="v-progress-linear__indeterminate long"
                        />
                        <div
                          class="v-progress-linear__indeterminate short"
                        />
                        
                      </div>
                    </transition-stub>
                    <!---->
                  </div>
                </div>
                <div
                  class="v-field__prepend-inner"
                >
                  <i
                    aria-hidden="true"
                    class="mdi-magnify mdi v-icon notranslate v-theme--light v-icon--size-default"
                  />
                  <!---->
                </div>
                <div
                  class="v-field__field"
                  data-no-activator=""
                >
                  <label
                    aria-hidden="true"
                    class="v-label v-field-label v-field-label--floating"
                    for="input-0"
                  >
                    <!---->
                    
                    Buscar personaje
                    
                  </label>
                  <label
                    class="v-label v-field-label"
                    for="input-0"
                  >
                    <!---->
                    
                    Buscar personaje
                    
                  </label>
                  
                  
                  <!---->
                  <input
                    aria-describedby="input-0-messages"
                    class="v-field__input"
                    id="input-0"
                    size="1"
                    type="text"
                  />
                  <!---->
                  
                  
                </div>
                <transition-stub
                  appear="false"
                  css="true"
                  name="expand-x-transition"
                  persisted="false"
                >
                  <div
                    class="v-field__clearable"
                    style="display: none;"
                  >
                    
                    <i
                      aria-hidden="false"
                      aria-label="Clear Buscar personaje"
                      class="mdi-close-circle mdi v-icon notranslate v-theme--light v-icon--size-default v-icon--clickable"
                      role="button"
                      tabindex="0"
                    />
                    
                  </div>
                </transition-stub>
                <div
                  class="v-field__append-inner"
                >
                  <!---->
                  <!---->
                </div>
                <div
                  class="v-field__outline"
                >
                  <!---->
                  <!---->
                </div>
              </div>
              
            </div>
            <!---->
            <div
              class="v-input__details"
            >
              <transition-group-stub
                appear="false"
                aria-live="polite"
                class="v-messages"
                css="true"
                id="input-0-messages"
                name="slide-y-transition"
                persisted="false"
                role="alert"
                tag="div"
              >
                <!---->
              </transition-group-stub>
              <!---->
            </div>
          </div>
        </div>
        <div
          class="v-col-sm-10 v-col-md-3 v-col-12"
        >
          <div
            class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-theme--light v-locale--is-ltr v-text-field v-select v-select--single"
          >
            <!---->
            <div
              class="v-input__control"
            >
              
              <div
                class="v-field v-field--appended v-field--center-affix v-field--prepended v-field--variant-filled v-theme--light v-locale--is-ltr"
                role="combobox"
              >
                <div
                  class="v-field__overlay"
                />
                <div
                  class="v-field__loader"
                >
                  <div
                    aria-hidden="true"
                    aria-valuemax="100"
                    aria-valuemin="0"
                    class="v-progress-linear v-theme--light v-locale--is-ltr"
                    role="progressbar"
                    style="top: 0px; height: 0px; --v-progress-linear-height: 2px;"
                  >
                    <!---->
                    <div
                      class="v-progress-linear__background"
                    />
                    <div
                      class="v-progress-linear__buffer"
                      style="width: 0%;"
                    />
                    <transition-stub
                      appear="false"
                      css="true"
                      name="fade-transition"
                      persisted="false"
                    >
                      <div
                        class="v-progress-linear__indeterminate"
                      >
                        
                        <div
                          class="v-progress-linear__indeterminate long"
                        />
                        <div
                          class="v-progress-linear__indeterminate short"
                        />
                        
                      </div>
                    </transition-stub>
                    <!---->
                  </div>
                </div>
                <div
                  class="v-field__prepend-inner"
                >
                  <i
                    aria-hidden="true"
                    class="mdi-filter mdi v-icon notranslate v-theme--light v-icon--size-default"
                  />
                  <!---->
                </div>
                <div
                  class="v-field__field"
                  data-no-activator=""
                >
                  <label
                    aria-hidden="true"
                    class="v-label v-field-label v-field-label--floating"
                    for="input-2"
                  >
                    <!---->
                    
                    Comics
                    
                  </label>
                  <label
                    class="v-label v-field-label"
                    for="input-2"
                  >
                    <!---->
                    
                    Comics
                    
                  </label>
                  
                  
                  <!---->
                  <div
                    class="v-field__input"
                    data-no-activator=""
                  >
                    
                    
                    
                    <!---->
                    <!---->
                    
                    
                    
                    
                    
                    <input
                      aria-describedby="input-2-messages"
                      aria-label="Open"
                      id="input-2"
                      inputmode="none"
                      size="1"
                      title="Open"
                      type="text"
                      value=""
                    />
                  </div>
                  <!---->
                  
                  
                </div>
                <transition-stub
                  appear="false"
                  css="true"
                  name="expand-x-transition"
                  persisted="false"
                >
                  <div
                    class="v-field__clearable"
                    style="display: none;"
                  >
                    
                    <i
                      aria-hidden="false"
                      aria-label="Clear Comics"
                      class="mdi-close-circle mdi v-icon notranslate v-theme--light v-icon--size-default v-icon--clickable"
                      role="button"
                      tabindex="0"
                    />
                    
                  </div>
                </transition-stub>
                <div
                  class="v-field__append-inner"
                >
                  
                  
                  <!---->
                  <i
                    aria-hidden="true"
                    class="mdi-menu-down mdi v-icon notranslate v-theme--light v-icon--size-default v-select__menu-icon"
                  />
                  
                  
                  <!---->
                </div>
                <div
                  class="v-field__outline"
                >
                  <!---->
                  <!---->
                </div>
              </div>
              
            </div>
            <!---->
            <div
              class="v-input__details"
            >
              <transition-group-stub
                appear="false"
                aria-live="polite"
                class="v-messages"
                css="true"
                id="input-2-messages"
                name="slide-y-transition"
                persisted="false"
                role="alert"
                tag="div"
              >
                <!---->
              </transition-group-stub>
              <!---->
            </div>
          </div>
        </div>
        <div
          class="v-col-sm-10 v-col-md-3 v-col-12"
        >
          <div
            class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-theme--light v-locale--is-ltr v-text-field v-select v-select--single"
          >
            <!---->
            <div
              class="v-input__control"
            >
              
              <div
                class="v-field v-field--appended v-field--center-affix v-field--prepended v-field--variant-filled v-theme--light v-locale--is-ltr"
                role="combobox"
              >
                <div
                  class="v-field__overlay"
                />
                <div
                  class="v-field__loader"
                >
                  <div
                    aria-hidden="true"
                    aria-valuemax="100"
                    aria-valuemin="0"
                    class="v-progress-linear v-theme--light v-locale--is-ltr"
                    role="progressbar"
                    style="top: 0px; height: 0px; --v-progress-linear-height: 2px;"
                  >
                    <!---->
                    <div
                      class="v-progress-linear__background"
                    />
                    <div
                      class="v-progress-linear__buffer"
                      style="width: 0%;"
                    />
                    <transition-stub
                      appear="false"
                      css="true"
                      name="fade-transition"
                      persisted="false"
                    >
                      <div
                        class="v-progress-linear__indeterminate"
                      >
                        
                        <div
                          class="v-progress-linear__indeterminate long"
                        />
                        <div
                          class="v-progress-linear__indeterminate short"
                        />
                        
                      </div>
                    </transition-stub>
                    <!---->
                  </div>
                </div>
                <div
                  class="v-field__prepend-inner"
                >
                  <i
                    aria-hidden="true"
                    class="mdi-filter mdi v-icon notranslate v-theme--light v-icon--size-default"
                  />
                  <!---->
                </div>
                <div
                  class="v-field__field"
                  data-no-activator=""
                >
                  <label
                    aria-hidden="true"
                    class="v-label v-field-label v-field-label--floating"
                    for="input-5"
                  >
                    <!---->
                    
                    Series
                    
                  </label>
                  <label
                    class="v-label v-field-label"
                    for="input-5"
                  >
                    <!---->
                    
                    Series
                    
                  </label>
                  
                  
                  <!---->
                  <div
                    class="v-field__input"
                    data-no-activator=""
                  >
                    
                    
                    
                    <!---->
                    <!---->
                    
                    
                    
                    
                    
                    <input
                      aria-describedby="input-5-messages"
                      aria-label="Open"
                      id="input-5"
                      inputmode="none"
                      size="1"
                      title="Open"
                      type="text"
                      value=""
                    />
                  </div>
                  <!---->
                  
                  
                </div>
                <transition-stub
                  appear="false"
                  css="true"
                  name="expand-x-transition"
                  persisted="false"
                >
                  <div
                    class="v-field__clearable"
                    style="display: none;"
                  >
                    
                    <i
                      aria-hidden="false"
                      aria-label="Clear Series"
                      class="mdi-close-circle mdi v-icon notranslate v-theme--light v-icon--size-default v-icon--clickable"
                      role="button"
                      tabindex="0"
                    />
                    
                  </div>
                </transition-stub>
                <div
                  class="v-field__append-inner"
                >
                  
                  
                  <!---->
                  <i
                    aria-hidden="true"
                    class="mdi-menu-down mdi v-icon notranslate v-theme--light v-icon--size-default v-select__menu-icon"
                  />
                  
                  
                  <!---->
                </div>
                <div
                  class="v-field__outline"
                >
                  <!---->
                  <!---->
                </div>
              </div>
              
            </div>
            <!---->
            <div
              class="v-input__details"
            >
              <transition-group-stub
                appear="false"
                aria-live="polite"
                class="v-messages"
                css="true"
                id="input-5-messages"
                name="slide-y-transition"
                persisted="false"
                role="alert"
                tag="div"
              >
                <!---->
              </transition-group-stub>
              <!---->
            </div>
          </div>
        </div>
      </div>
    </div>
    <!--v-if-->
    <div
      class="v-row justify-center"
    >
      <div
        class="v-col v-col-10"
      >
        <div
          class="v-alert v-theme--light bg-info v-alert--density-default v-alert--variant-flat"
          role="alert"
        >
          
          <!---->
          <span
            class="v-alert__underlay"
          />
          
          <!---->
          <div
            class="v-alert__prepend"
          >
            <i
              aria-hidden="true"
              class="mdi-information mdi v-icon notranslate v-theme--light"
              density="default"
              style="font-size: 28px; height: 28px; width: 28px;"
            />
          </div>
          <div
            class="v-alert__content"
          >
            <!---->
            <!---->
            
            No existe información para los filtros seleccionados.
            
          </div>
          <!---->
          <!---->
        </div>
      </div>
    </div>
    <!--v-if-->
    
    <!---->
    <!---->
    
    <div
      class="text-center"
    >
      <div
        class="v-container v-locale--is-ltr"
      >
        <div
          class="v-row justify-center"
        >
          <div
            class="v-col v-col-8"
          >
            <div
              class="v-container v-locale--is-ltr max-width"
            >
              <nav
                aria-label="Pagination Navigation"
                class="v-pagination v-theme--light my-4"
                data-test="v-pagination-root"
                role="navigation"
              >
                <ul
                  class="v-pagination__list"
                >
                  <!---->
                  <li
                    class="v-pagination__prev"
                    data-test="v-pagination-prev"
                  >
                    <button
                      aria-disabled="true"
                      aria-label="Previous page"
                      class="v-btn v-btn--disabled v-btn--icon v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-text"
                      disabled=""
                      type="button"
                    >
                      
                      <span
                        class="v-btn__overlay"
                      />
                      <span
                        class="v-btn__underlay"
                      />
                      
                      <!---->
                      <span
                        class="v-btn__content"
                        data-no-activator=""
                      >
                        <i
                          aria-hidden="true"
                          class="mdi-chevron-left mdi v-icon notranslate v-theme--light v-icon--size-default"
                        />
                      </span>
                      <!---->
                      <!---->
                    </button>
                  </li>
                  
                  <li
                    class="v-pagination__item v-pagination__item--is-active"
                    data-test="v-pagination-item"
                  >
                    <button
                      aria-current="true"
                      aria-label="Page 1, Current page"
                      class="v-btn v-btn--disabled v-btn--icon v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-text"
                      disabled=""
                      ellipsis="false"
                      type="button"
                    >
                      
                      <span
                        class="v-btn__overlay"
                      />
                      <span
                        class="v-btn__underlay"
                      />
                      
                      <!---->
                      <span
                        class="v-btn__content"
                        data-no-activator=""
                      >
                        
                        
                        1
                        
                        
                      </span>
                      <!---->
                      <!---->
                    </button>
                  </li>
                  
                  <li
                    class="v-pagination__next"
                    data-test="v-pagination-next"
                  >
                    <button
                      aria-disabled="true"
                      aria-label="Next page"
                      class="v-btn v-btn--disabled v-btn--icon v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-text"
                      disabled=""
                      type="button"
                    >
                      
                      <span
                        class="v-btn__overlay"
                      />
                      <span
                        class="v-btn__underlay"
                      />
                      
                      <!---->
                      <span
                        class="v-btn__content"
                        data-no-activator=""
                      >
                        <i
                          aria-hidden="true"
                          class="mdi-chevron-right mdi v-icon notranslate v-theme--light v-icon--size-default"
                        />
                      </span>
                      <!---->
                      <!---->
                    </button>
                  </li>
                  <!---->
                </ul>
              </nav>
            </div>
          </div>
        </div>
      </div>
    </div>
    
  </div>
</div>
`;
